libbpf-tools/cachestat.c

```diff
@@ -0,0 +1,208 @@
+"Count cache kernel function calls.\n"
+"\n"
+"USAGE: cachestat [--help] [-T] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    cachestat           # shows hits and misses to the file system page cache"
+"    cachestat -T        # include timestamps"
+"    cachestat 1 10      # print 1 second summaries, 10 times\n";
+
+static const struct argp_option opts[] = {
+	{ "timestamp", 'T', NULL, 0, "Print timestamp" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static int get_meminfo(__u64 *buffers, __u64 *cached)
+{
+	__u64 tmp;
+	FILE *f;
+
+	f = fopen("/proc/meminfo", "r");
+	if (!f)
+		return -1;
+	if (fscanf(f, "MemTotal: %llu kB MemFree: %llu kB MemAvailable: %llu "
+			"kB Buffers: %llu kB Cached: %llu kB", &tmp, &tmp, &tmp,
+			buffers, cached) != 5) {
+		fclose(f);
+		return -1;
+	}
+	fclose(f);
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct counter counter, zero = {};
+	struct cachestat_bpf *obj;
+	__u64 buffers, cached;
+	struct tm *tm;
+	float ratio;
+	char ts[32];
+	__s64 hits;
+	time_t t;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = cachestat_bpf__open_and_load();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	err = cachestat_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	signal(SIGINT, sig_handler);
+
+	if (env.timestamp)
+		printf("%-8s ", "TIME");
+	printf("%8s %8s %8s %8s %12s %10s\n", "HITS", "MISSES", "DIRTIES",
+		"HITRATIO", "BUFFERS_MB", "CACHED_MB");
+
+	while (1) {
+		sleep(env.interval);
+
+		counter = obj->bss->counter;
+		obj->bss->counter = zero;
```

Comment:
you get no benefits from having `struct counter`. It neither saves code, nor gives you atomicity. Let's drop it (and cachestat.h as well) and just use 3 straight global variables. On user-space side, it's probably a good idea to use atomic swap to zero to not miss any counters. Alternatively, you can do always increasing counters and just remember the last state here, then calculate differences for the period. Makes sense?

Reply:
Fixed with atomic swap to zero. thanks!